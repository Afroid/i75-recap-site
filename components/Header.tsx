import { useState } from "react";
import HeaderLogo from "@/components/HeaderLogo";
import DesktopNavbar from "@/components/DesktopNavbar";
import MobileDrawer from "@/components/MobileDrawer";
import { TestIds } from "@/lib/testIds";

/**
 * Site-wide header component.
 *
 * - Shows a logo and hamburger menu on mobile
 * - Shows desktop navigation on larger screens
 * - Manages mobile drawer toggle state
 */
export default function Header() {
  // This is for the mobile drawer, whether it's open or not
  const [isOpen, setIsOpen] = useState(false);

  // Toggle open and close for mobile drawer
  const toggleDrawer = () => setIsOpen((prev) => !prev);
  // Close for mobile drawer
  const closeDrawer = () => setIsOpen(false);

  return (
    <>
      {/* Main Header */}
      <header
        data-testid={TestIds.HEADER_SITE_HEADER}
        className="fixed top-0 left-0 w-full z-50 bg-global-nav text-white shadow-md h-12"
      >
        <div className="max-w-7xl mx-auto flex items-center justify-between h-full">

          {/* Mobile Hamburger and Logo*/}
          <div
            data-testid={TestIds.HEADER_MOBILE_HEADER}
            className="flex items-center space-x-4 lg:hidden bg-green-500 h-full pl-4"
          >
            <button
              data-testid={TestIds.HEADER_HAMBURGER_BUTTON}
              onClick={toggleDrawer}
              aria-label="Toggle menu"
              className="relative w-8 h-8 flex items-center justify-center"
            >
              {/* Top Bar */}
              <div
                className={`
                  absolute w-8 h-0.5 bg-white transform
                  transition-all duration-300 ease-in-out
                  ${isOpen ? "rotate-45" : "-translate-y-2"}
                `}
              />
              {/* Middle Bar */}
              <div
                className={`absolute w-8 h-0.5 bg-white transition-all duration-300 ease-in-out
                  ${isOpen ? "opacity-0" : ""}
                `}
              />
              {/* Bottom Bar */}
              <div
                className={`
                  absolute w-8 h-0.5 bg-white transform
                  transition-all duration-300 ease-in-out
                  ${isOpen ? "-rotate-45" : "translate-y-2"}
                `}
              />
            </button>

            {/* Mobile Logo (clicking returns home) */}
            <HeaderLogo data-testid={TestIds.HEADER_MOBILE_LOGO} />
          </div>

          {/* Desktop HeaderLogo (hidden on mobile) */}
          <div data-testid={TestIds.HEADER_DESKTOP_LOGO} className="hidden lg:flex">
            <HeaderLogo />
          </div>

          {/* Desktop Nav (hidden on mobile) */}
          <nav
            data-testid={TestIds.HEADER_DESKTOP_NAV}
            className="hidden lg:flex items-center flex-grow ml-8">
            <DesktopNavbar />
          </nav>

        </div>

        {/* Mobile Drawer (slides out) */}
        <MobileDrawer
          data-testid={TestIds.HEADER_MOBILE_DRAWER}
          isOpen={isOpen}
          onClose={closeDrawer}
        />
      </header>

    </>
  );
}
