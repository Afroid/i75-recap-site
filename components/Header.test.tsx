import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";

// Stub dependencies
jest.mock("@/components/HeaderLogo", () => ({
  __esModule: true,
  default: () => <div data-testid="stubbed-header-logo">Logo</div>,
}));
jest.mock("@/components/DesktopNavbar", () => ({
  __esModule: true,
  default: () => <nav data-testid="stubbed-desktop-nav">DesktopNav</nav>,
}));
jest.mock("@/components/MobileDrawer", () => ({
  __esModule: true,
  default: ({ isOpen, onClose }: { isOpen: boolean; onClose: () => void }) => (
    <div
      data-testid="mobile-drawer"
      data-is-open={isOpen}
      onClick={onClose}
    >
      Drawer({isOpen ? "open" : "closed"})
    </div>
  ),
}));

/**
 * Importing after the mocks ensures the real copies of those dependencies don't get loaded.
 * This ensures any imports from next/head or whatever get routed to the stubbed versions, rather
 * than real implementation. This makes sure jest.mock calls will have effect.
 */
// Now import Header (after the mocks)
import Header from "./Header";

describe("Header", () => {
  it("renders all major sections", () => {
    render(<Header />);

    const logos = screen.getAllByTestId("stubbed-header-logo");

    // Asserts there should be two instances (mobile and desktop)
    // because we aren't accounting for screen size
    expect(logos).toHaveLength(2);

    // Asserts presence of logos, desktop nav, hamburger button, drawer container
    expect(logos[0]).toBeInTheDocument();
    expect(logos[1]).toBeInTheDocument();
    expect(screen.getByTestId("stubbed-desktop-nav")).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "Toggle menu" })).toBeInTheDocument();
    expect(screen.getByTestId("mobile-drawer")).toHaveAttribute("data-is-open", "false");
  });

  it("toggles mobile drawer open and closed", () => {
    render(<Header />);
    const toggle = screen.getByRole("button", { name: "Toggle menu" });
    const drawer = screen.getByTestId("mobile-drawer");

    // Initial state = closed
    expect(drawer).toHaveAttribute("data-is-open", "false");

    // Click on hamburger menu to change the state and open the mobile drawer
    fireEvent.click(toggle);
    expect(drawer).toHaveAttribute("data-is-open", "true");

    // Click on drawer to trigger onClose
    fireEvent.click(drawer);
    expect(drawer).toHaveAttribute("data-is-open", "false");
  });

  it("matches the snapshot", () => {
    const { asFragment } = render(<Header />);
    expect(asFragment()).toMatchSnapshot();
  });
});
