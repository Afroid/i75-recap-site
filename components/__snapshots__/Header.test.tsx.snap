// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header matches the snapshot 1`] = `
<DocumentFragment>
  <header
    class="fixed top-0 left-0 w-full z-50 bg-global-nav text-white shadow-md h-12"
    data-testid="site-header"
  >
    <div
      class="max-w-7xl mx-auto flex items-center justify-between h-full"
    >
      <div
        class="flex items-center space-x-4 lg:hidden bg-green-500 h-full pl-4"
        data-testid="mobile-header"
      >
        <button
          aria-label="Toggle menu"
          class="relative w-8 h-8 flex items-center justify-center"
          data-testid="header-hamburger-button"
        >
          <div
            class="
                  absolute w-8 h-0.5 bg-white transform
                  transition-all duration-300 ease-in-out
                  -translate-y-2
                "
          />
          <div
            class="absolute w-8 h-0.5 bg-white transition-all duration-300 ease-in-out
                  
                "
          />
          <div
            class="
                  absolute w-8 h-0.5 bg-white transform
                  transition-all duration-300 ease-in-out
                  translate-y-2
                "
          />
        </button>
        <div
          data-testid="stubbed-header-logo"
        >
          Logo
        </div>
      </div>
      <div
        class="hidden lg:flex"
        data-testid="header-desktop-logo"
      >
        <div
          data-testid="stubbed-header-logo"
        >
          Logo
        </div>
      </div>
      <nav
        class="hidden lg:flex items-center flex-grow ml-8"
        data-testid="header-desktop-nav"
      >
        <nav
          data-testid="stubbed-desktop-nav"
        >
          DesktopNav
        </nav>
      </nav>
    </div>
    <div
      data-is-open="false"
      data-testid="mobile-drawer"
    >
      Drawer(closed)
    </div>
  </header>
</DocumentFragment>
`;
